test/modules/packages/linearalgebra/correctness/correctness.chpl

```diff
@@ -0,0 +1,435 @@
+  }
+
+
+  /* Dimensions */
+  {
+    var M = Matrix(3, 3);
+    assertEqual(M.domain, MDom, "Matrix(3, 3)");
+  }
+
+  /* Domain */
+  {
+    var M = Matrix(MDom);
+    assertEqual(M.domain, MDom, "Matrix(MDom)");
+  }
+
+  /* Array */
+  {
+    var A: [MDom] real;
+    var M = Matrix(A);
+    assertEqual(M.domain, MDom, "Matrix(A)");
+  }
+
+  /* Variadic arguments of vectors */
+  {
+    var M = Matrix([1,2,3],
+                   [4,5,6],
+                   [7,8,9]);
+    assertEqual(M.domain, MDom, "Matrix([1,2,3], [4,5,6], [7,8,9]).domain");
+    assertEqual(M[1,1], 5, "Matrix([1,2,3], [4,5,6], [7,8,9])[1,1]");
+  }
+
+  /* Identity - Dimensions */
+  {
+    proc checkIdentity(m, n, type t = real) {
+      var Identity: [0..#m, 0..#n] t;
+      const count = min(m, n);
+      for i in 0..#count {
+        Identity[i, i] = 1: t;
+      }
+
+      var I = eye(m, n, eltType=t);
+      assertEqual(I, Identity,
+          "checkIdentity(%i, %i, %s)".format(m, n, t:string));
+    }
+
+    checkIdentity(3, 3);
+    checkIdentity(3, 4);
+    checkIdentity(4, 3);
+    checkIdentity(3, 3, complex);
+  }
+
+  {
+    var Identity: [MDom] real;
+    for i in MDom.dim(1) do Identity[i, i] = 1.0;
+
+    var I = eye(MDom);
+    assertEqual(I, Identity, "eye(D)");
+  }
+}
+
+//
+// Matrix operations
+//
+
+
+{
+  /* transpose - vector */
+  {
+    var v = Vector([1,2,3]);
+    var vT = transpose(v);
+
+    assertEqual(v, vT, "transpose(v)");
+    assertEqual(v.shape, vT.shape, "transpose(v).shape");
+    assertEqual(v.T, vT, "v.T");
+  }
+
+  /* transpose - Nx1 , 1xN vectors */
+  {
+    var v31 = Matrix(3, 1);
+    var v13 = Matrix(1, 3);
+
+    v31 = 1;
+    v13 = 1;
+
+    assertEqual(v13, v31.T, "Matrix(3, 1).T");
+    assertEqual(v31, v13.T, "Matrix(1, 3).T");
+  }
+
+  /* transpose - square matrix */
+  {
+
+    var A =  Matrix([1.0,2.0,3.0],
+                    [4.0,5.0,6.0],
+                    [7.0,8.0,9.0]);
+    var AT = Matrix([1.0,4.0,7.0],
+                    [2.0,5.0,8.0],
+                    [3.0,6.0,9.0]);
+
+    assertEqual(transpose(A).shape, AT.shape, "transpose(A).shape");
+    assertEqual(transpose(A), AT, "transpose(A)");
+    assertEqual(A.T, AT, "A.T");
+  }
+
+  /* transpose - non-square matrix */
+  {
+    var A =  Matrix([1,2,3],
+                    [4,5,6]);
+
+    var AT = Matrix([1,4],
+                    [2,5],
+                    [3,6]);
+
+    assertEqual(transpose(A).shape, AT.shape, "transpose(A).shape");
+    assertEqual(transpose(A), AT, "transpose(A)");
+    assertEqual(A.T, AT, "A.T");
+  }
+}
+
+/* dot - calls matMult && inner*/
+{
+
+  proc test_dot(type t) {
+    var M = Matrix([1,2,3],
+                   [4,5,6],
+                   [7,8,9],
+                   eltType=t);
+    var I = eye(M.domain, eltType=t);
+
+    var v = Vector(1,2,3, eltType=t);
+
+    // matrix-matrix
+    var MI = dot(M, I);
+    assertEqual(MI, M, "dot(M, I)");
+
+    // vector-matrix && matrix-vector
+    var Iv = dot(I, v);
```

Comment:
I think we should do this test with M and not I. The identity is a little too special when it comes to linear algebra (eg. it would not have caught the bug in the gemv transpose).